### PR TITLE
Improve performance of Cellular Automaton for regional tracking at HLT

### PR DESCRIPTION
**Pull request by JanFSchulte**
So far, for each tracking region at the HLT, a new CA Graph structure is created. This update intends to improve performance by re-using the data structure and re-filling it for each of the regions. For this, some code is rearranged, but the algorithm itself is not changed at all. 

Some details and performance tests can be found here: https://cernbox.cern.ch/index.php/s/UWvE7a7itx6VGBs

The executive summary is:
- Checked physics results both offline and online -> 100% identical 
- Checked timing of modules: Modules using regional tracking speed up between 3 and 40%. The gain strongly depends on the size of the regions, for many small regions the reduction in overhead naturally matters more than for few large ones.

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
Thanks Jan. Could you consider replacing "efficiency" with e.g. "performance" (or "speed") in PR title and description? Just to avoid a possible misconception that physics performance would be altered (which was my first interpretation of "improve efficiency").

**Comment by JanFSchulte** [comment]
Very good point. Fixed.

**Comment by JanFSchulte** [comment]
Could someone please review this? 

I have performed additional performance studies using igprof on a reduced HLT menu running a couple of TkMu paths. Shows 14% improvement for the quadruplet building and 6% for the overall event.

**Comment by slava77** [comment]
On 2/8/17 5:05 AM, Jan-Frederik Schulte wrote:
> Could someone please review this?
> 
> I have performed additional performance studies using igprof on a
> reduced HLT menu running a couple of TkMu paths. Shows 14% improvement
> for the quadruplet building and 6% for the overall event.

It's next on my list for review.
Expect some feedback today.

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletEDProducerT.cc [hunk 1]
```diff
@@ -86,21 +86,19 @@ void CAHitNtupletEDProducerT<T_Generator>::produce(edm::Event& iEvent, const edm
   generator_.initEvent(iEvent, iSetup);
 
   LogDebug("CAHitNtupletEDProducer") << "Creating ntuplets for " << regionDoublets.regionSize() << " regions, and " << regionDoublets.layerPairsSize() << " layer pairs";
-
-  typename T_Generator::ResultType ntuplets;
-  ntuplets.reserve(localRA_.upper());
-
+  std::vector<typename T_Generator::ResultType> ntuplets;
+  ntuplets.resize(regionDoublets.regionSize());
+  for(auto ntuplet : ntuplets)  ntuplet.reserve(localRA_.upper());
```
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletEDProducerT.cc]: maybe use ``` auto& ``` or ```auto&& ```
 to be a bit more obvious that it's not a temp copy that's modified.

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc [hunk 2]
```diff
@@ -216,21 +247,194 @@ void CAHitQuadrupletGenerator::hitNtuplets(const IntermediateHitDoublets::Region
...
+      			}
+    		}
+
+    		float chi2 = std::numeric_limits<float>::quiet_NaN();
+    		// TODO: Do we have any use case to not use bending correction?
+    		if (useBendingCorrection)
+   		{
+      		// Following PixelFitterByConformalMappingAndLine
+      		const float simpleCot = ( gps.back().z() - gps.front().z() ) / (gps.back().perp() - gps.front().perp() );
+     		const float pt = 1.f / PixelRecoUtilities::inversePt(abscurv, es);
```
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc]: some indentation to offset the code in the ```if (useBendingCorrection)``` scope would be useful for readability

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletEDProducerT.cc (hunk 1 above)
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletEDProducerT.cc]: I added some printouts and It is clear that this loop is now doing operations on the copies.
Please fix.

**Comment by JanFSchulte** [comment]
Thanks for reviewing, @slava77 . The new commit should address both your comments.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
I ran 10224 (ttbar PU35 with CA era Run2_2017_trackingPhase1CA)
compared to commit 48d82fe, the latest one has a small increase in CPU

```
"detachedQuad\|detachedTriplet\|highPtTriplet\|initialStep\|lowPtQuad\|lowPtTriplet" |  grep -v Tracks
   +0.119492      +0.01 %        48.47 ms/ev ->        54.63 ms/ev initialStepHitQuadruplets
   +0.117181      +0.01 %        48.58 ms/ev ->        54.63 ms/ev initialStepHitQuadrupletsPreSplitting
   +0.096095      +0.02 %        76.45 ms/ev ->        84.17 ms/ev detachedQuadStepHitQuadruplets
   +0.085959      +0.03 %       144.01 ms/ev ->       156.95 ms/ev lowPtQuadStepHitQuadruplets
   +0.081517      +0.01 %        29.49 ms/ev ->        31.99 ms/ev highPtTripletStepHitTriplets
   +0.080369      +0.02 %        99.79 ms/ev ->       108.15 ms/ev lowPtTripletStepHitTriplets
   +0.074207      +0.01 %        49.95 ms/ev ->        53.80 ms/ev detachedTripletStepHitTriplets
Todal: 496.74 => 544.32	 delta 47.58
```

@JanFSchulte please check your HLT tests so that timing improvements remain
and were not blown away by the latest update which is just a one symbol (&) added in ```for(auto& ntuplet : ntuplets)  ntuplet.reserve(localRA_.upper()); ```

**Comment by slava77** [comment]
@JanFSchulte 
did you have a chance to follow up on my comment in https://github.com/cms-sw/cmssw/pull/17373#issuecomment-279016363
?
I'm holding off signing this PR until it's cleared.

**Comment by JanFSchulte** [comment]
@slava77 What I can tell you so far is that I had checked the change in performance using igprof and the numbers for the produce() method did not change. I also asked @makortel for feedback, as he knows the code much better. He suggested a couple of checks that I will run now. Expect a conclusion later today.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/17373	77f9a41
- code changes are in line with the description and the follow up review; the change is not yet affecting the standard workflows
- jenkins tests pass and comparisons show no differences
- local test with Run2_2017_trackingPhase1CA era shows that there are no changes in results of step3 (step2/HLT is not modified by this era).

@JanFSchulte for me your check with igprof is good enough.

**Comment by davidlange6** [comment]
+1

**Pull request merged**